### ISSUE TITLE
Get `typst` packaged for Debian and other repos

### ISSUE THREAD

**Issue by tgross35** [typst/typst#3679]
### Description

In order to get `typst` installable via the Debian repos, all dependencies in `Cargo.lock` need to be available at https://qa.debian.org/developer.php?email=pkg-rust-maintainers%40alioth-lists.debian.net. The majority exist already, but from a cursory look maybe about 20% will need to be added if they can't be configured out.

This happens by adding dependencies in salsa via debcargo https://salsa.debian.org/rust-team/debcargo-conf which anyone can do. Once all the missing dependencies are in, Typst itself can be added.

There is an existing RTP at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1060888

There is also an incomplete https://github.com/jcbhmr/typst.deb, which I believe would be for a PPA. Being in the main repos is of course a better goal, if possible.

### Use Case

It would be an awesome convenience to be able to install Typst via `apt`. I know the Typst maintainers will likely not want to get involved, but I think it seems reasonable to have a "help wanted" tracking issue.

**Comment by extua** [comment]
To track some of those dependencies already in Debian, I went through the [Debian Rust Maintainers list](https://qa.debian.org/developer.php?email=pkg-rust-maintainers%40alioth-lists.debian.net) and checked off the crates already present.

- [ ] typst
- [ ] typst-cli
- [ ] typst-ide
- [ ] typst-macros
- [ ] typst-pdf
- [ ] typst-render
- [ ] typst-svg
- [ ] typst-syntax
- [ ] typst-timing
- [ ] typst-assets
- [ ] typst-dev-assets
- [ ] az
- [x] [base64](https://tracker.debian.org/pkg/rust-base64)
- [x] [bitflags](https://tracker.debian.org/pkg/rust-bitflags)
- [x] [bytemuck](https://tracker.debian.org/pkg/rust-bytemuck)
- [ ] chinese-number
- [x] [chrono](https://tracker.debian.org/pkg/rust-chrono)
- [x] [ciborium](https://tracker.debian.org/pkg/rust-ciborium)
- [x] [clap](https://tracker.debian.org/pkg/rust-clap)
- [x] [clap_complete](https://tracker.debian.org/pkg/rust-clap-complete)
- [x] [clap_mangen](https://tracker.debian.org/pkg/rust-clap-mangen)
- [x] [ctrlc](https://tracker.debian.org/pkg/rust-ctrlc)
- [x] [codespan-reporting](https://tracker.debian.org/pkg/rust-codespan-reporting)
- [ ] comemo
- [x] [csv](https://tracker.debian.org/pkg/rust-csv)
- [x] [dirs](https://tracker.debian.org/pkg/rust-dirs)
- [ ] ecow
- [ ] env_proxy
- [x] [flate2](https://tracker.debian.org/pkg/rust-flate2)
- [ ] fontdb
- [x] [fs_extra](https://tracker.debian.org/pkg/rust-fs-extra)
- [ ] hayagriva
- [x] [heck](https://tracker.debian.org/pkg/rust-heck)
- [ ] hypher
- [ ] icu_properties
- [ ] icu_provider
- [ ] icu_provider_adapters
- [ ] icu_provider_blob
- [ ] icu_segmenter
- [ ] if_chain
- [x] [image](https://tracker.debian.org/pkg/rust-image)
- [x] [indexmap](https://tracker.debian.org/pkg/rust-indexmap)
- [x] [kamadak-exif](https://tracker.debian.org/pkg/rust-kamadak-exif)
- [x] [kurbo](https://tracker.debian.org/pkg/rust-kurbo)
- [ ] libfuzzer-sys
- [ ] lipsum
- [x] [log](https://tracker.debian.org/pkg/rust-log)
- [x] [miniz_oxide](https://tracker.debian.org/pkg/rust-miniz-oxide)
- [x] [native-tls](https://tracker.debian.org/pkg/rust-native-tls)
- [x] [notify](https://tracker.debian.org/pkg/rust-notify)
- [x] [once_cell](https://tracker.debian.org/pkg/rust-once-cell)
- [x] [open](https://tracker.debian.org/pkg/rust-open)
- [x] [openssl](https://tracker.debian.org/pkg/rust-openssl)
- [ ] oxipng
- [x] [palette](https://tracker.debian.org/pkg/rust-palette)
- [x] [parking_lot](https://tracker.debian.org/pkg/rust-parking-lot)
- [x] [pathdiff](https://tracker.debian.org/pkg/rust-pathdiff)
- [ ] pdf-writer
- [x] [phf](https://tracker.debian.org/pkg/rust-phf)
- [ ] pixglyph
- [x] [png](https://tracker.debian.org/pkg/rust-png)
- [x] [portable-atomic](https://tracker.debian.org/pkg/rust-portable-atomic)
- [x] [proc-macro2](https://tracker.debian.org/pkg/rust-proc-macro2)
- [x] [pulldown-cmark](https://tracker.debian.org/pkg/rust-pulldown-cmark)
- [x] [quote](https://tracker.debian.org/pkg/rust-quote)
- [ ] qcms
- [x] [rayon](https://tracker.debian.org/pkg/rust-rayon)
- [x] [regex](https://tracker.debian.org/pkg/rust-regex)
- [ ] resvg
- [x] [roxmltree](https://tracker.debian.org/pkg/rust-roxmltree)
- [ ] rustybuzz
- [x] [same-file](https://tracker.debian.org/pkg/rust-same-file)
- [ ] self-replace
- [x] [semver](https://tracker.debian.org/pkg/rust-semver)
- [x] [serde](https://tracker.debian.org/pkg/rust-serde)
- [x] [serde_json](https://tracker.debian.org/pkg/rust-serde-json)
- [x] [serde_yaml](https://tracker.debian.org/pkg/rust-serde-yaml)
- [x] [shell-escape](https://tracker.debian.org/pkg/rust-shell-escape)
- [x] [siphasher](https://tracker.debian.org/pkg/rust-siphasher)
- [x] [smallvec](https://tracker.debian.org/pkg/rust-smallvec)
- [x] [stacker](https://tracker.debian.org/pkg/rust-stacker)
- [ ] subsetter
- [ ] svg2pdf
- [x] [syn](https://tracker.debian.org/pkg/rust-syn)
- [x] [syntect](https://tracker.debian.org/pkg/rust-syntect)
- [x] [tar](https://tracker.debian.org/pkg/rust-tar)
- [x] [tempfile](https://tracker.debian.org/pkg/rust-tempfile)
- [x] [time](https://tracker.debian.org/pkg/rust-time)
- [x] [tiny-skia](https://tracker.debian.org/pkg/rust-tiny-skia)
- [x] [toml](https://tracker.debian.org/pkg/rust-toml)
- [x] [ttf-parser](https://tracker.debian.org/pkg/rust-ttf-parser)
- [ ] two-face
- [x] [typed-arena](https://tracker.debian.org/pkg/rust-typed-arena)
- [x] [unicode-bidi](https://tracker.debian.org/pkg/rust-unicode-bidi)
- [x] [unicode-ident](https://tracker.debian.org/pkg/rust-unicode-ident)
- [ ] unicode-math-class
- [x] [unicode-properties](https://tracker.debian.org/pkg/rust-unicode-properties)
- [x] [unicode-script](https://tracker.debian.org/pkg/rust-unicode-script)
- [x] [unicode-segmentation](https://tracker.debian.org/pkg/rust-unicode-segmentation)
- [ ] unscanny
- [ ] ureq
- [ ] usvg
- [x] [walkdir](https://tracker.debian.org/pkg/rust-walkdir)
- [ ] wasmi
- [x] [xmlparser](https://tracker.debian.org/pkg/rust-xmlparser)
- [x] [xmlwriter](https://tracker.debian.org/pkg/rust-xmlwriter)
- [ ] xmp-writer
- [x] [xz2](https://tracker.debian.org/pkg/rust-xz2)
- [ ] yaml-front-matter
- [x] [zip](https://tracker.debian.org/pkg/rust-zip)

**Comment by boredsquirrel** [comment]
I am wondering. Why are there no official binaries? rustup and cargo allow to keep all the rust libraries in the user dir, taking the load off the Distribution. These should work everywhere, and can simply be installed for the user.

```
# install rustup according to https://www.rust-lang.org/tools/install
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

# add it to your path, example for bash
echo "PATH=$PATH:$HOME/.cargo/bin" >> ~/.bashrc

cd ~
cargo init
cargo install typst
```

**Comment by laurmaedje** [comment]
@boredsquirrel We do have official binaries attached to each release, e.g.: https://github.com/typst/typst/releases/tag/v0.11.0

**Comment by boredsquirrel** [comment]
https://github.com/typst/typst/issues/3966

**Comment by rikhuijzer** [comment]
> It would be an awesome convenience to be able to install Typst via `apt`.

I don't think so. Most binaries on Debian are severely outdated due to the the way they package the distribution, so that makes it very inconvenient to package via them.

On a fully up-to-date Debian bookworm (bookworm is the newest version of Debian) system:

```sh
$ sudo apt-get install neovim

$ nvim --version
NVIM v0.7.2
```
Nvim is at 0.9.5 since Dec 30, 2023. This old v0.7.2 was released Jun 26, 2022. So the newest Debian is **one and a half year behind!**

The same occurs basically everywhere. `clang` is at 14.0.6 in Debian while it should be at 18.1.5. Clang 14.0.6 was released 24 Jun 2022. The newest version was released 17 April 2024, so they are almost two years behind.

It's inherent to Debian. That's why projects like Docker use a separate apt repository, which is also cumbersome. Instead, I second @boredsquirrel that using cargo or any other package manager than Debian is probably a better way to go.

**Comment by mquinson** [comment]
I agree that this is sort of an issue for Debian, not for typst itself. That being said, as a user, I prefer to have my software centrally managed by apt and dpkg than manually installed. Piping a script you got from wget into the shell is particularly scary to me. I understand that it's the Rust way to go and that's perfectly ok, but I really prefer not to do it at home.

This issue remains very precious even if it's mostly a downstream issue for Debian, not an upstream issue for Typst, because of the careful breakup of missing dependencies that still need to be packaged.

**Comment by rikhuijzer** [comment]
> Piping a script you got from wget into the shell is particularly scary to me. I understand that it's the Rust way to go and that's perfectly ok, but I really prefer not to do it at home.

With `cargo-binstall`, installing `typst` would not involve running arbitrary scripts. Instead, it could look like:
```sh
$ cargo binstall typst
```
assuming typst is set up to allow this. bininstall would use the official GitHub releases, so as long as nobody gains access to make releases, it should be safe.

Wget/curl into shell is not inherently less safe than `apt-get` by the way. For example
```sh
$ curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```
enforces HTTPS. In this case, the extra `curl` setting ensure that the data can only be downloaded from `sh.rustup.rs` and nobody can alter the data in-transit. So as long as the domain is trustworthy, it should be fine. Even Debian is susceptible to supply chain attacks as was shown by the XZ backdoor.

**Comment by ldegen** [comment]
> > It would be an awesome convenience to be able to install Typst via `apt`.
> 
> I don't think so. Most binaries on Debian are severely outdated due to the the way they package the distribution, so that makes it very inconvenient to package via them.

Agreed, from the developer/upstream perspective this can be really painful.

But from a user perspective having stuff packaged in the official repositories totally rocks.

Also, that "Debian is outdated" trope tickles me to no end.

First of all, it really depends on what you mean by outdated. If we are talking security-related patches, then *any* of the Debian releases is more likely to give you a more "up-to-date" binary then some random tarball that you downloaded half a year ago from some website you do not remember the name of. That's why I prefer having my software packages managed by the distribution whenever possible. Just one thing less I need to worry about. 
It's not at all about "how easy is it to install a package". Rather, it is about "how do I keep track of stuff afterwards".

Secondly, `Bookworm` is the name of the current *stable* release of Debian. Calling it the "newest" version is slightly misrepresenting things.  If having the latest upstream package versions is more important to you than stability, you should probably not be running a stable release anyway. For example, in both `Trixie` (*testing*) and `Sid` (*unstable*),  neovim is at 0.9.5 and clang at  16.0.

Sorry for the digression :heart:

**Comment by InAnYan** [comment]
This looks so strange, that such a serious app as Typst doesn't have Debian/Ubuntu versions...

**Comment by P-EB** [comment]
> This looks so strange, that such a serious app as Typst doesn't have Debian/Ubuntu versions...

This is not really strange when you get used to how distros works.

Rust, go, python etc are high level languages (I know rust also IS a low-level language), with therefore a tendancy to have a lot of small bricks one needs to bundle together.

Any top-level application therefore transitively depends on a LOT of things. As you see upwards, plenty deps are missing.

Packaging dependencies takes time. Volunteer-based time when it comes to Debian.

In the whole it's just that : lack of volunteer time.

Typst will end up in Debian. But we need to wait for the people in the rust team to package what is missing.

**Comment by bbolker** [comment]
I am super-sympathetic to the time constraints (and I don't, unfortunately, have the expertise and time to step up to this challenge), but I did want to offer a +1 to

> from a user perspective having stuff packaged in the official repositories totally rocks

**Comment by LS-KR** [comment]
Could this be a temporary solution?

https://github.com/LS-KR/typst-deb/ releases/tag/0.13.1

**Comment by eichin** [comment]
If I'm doing this right (clean Debian Testing install under podman, package names from the lockfile turned into debian package names, and the [Debian-specific tweaks](https://wiki.debian.org/Rust) to `config.toml` to repoint everything at `/usr/share/cargo/registry`... the only things `cargo fetch` complains about now are

* `typst-assets`
* `typst-dev-assets`
* `LaurenzV/krilla`
* `comemo`

I guess there are some deps hidden behind those - for example, if cargo can't fetch krilla, it doesn't expose that krilla also needs

* hayro-write
* once_cell
* pdf-writer
* skrifa
* subsetter
* xmp-writer
* yoke

but it looks like the 261 packages we do have directly are surprisingly close to enough! Kudos to everyone involved behind the scenes (mostly hidden behind the "Debian Rust Maintainers" team banner.)

**Comment by eichin** [comment]
Taking a different approach (adding a `[patch.crates-io]` section to `Cargo.toml` listing all of the local packages explicitly with `path = /usr/share/cargo/registry/pkg-version` entries for each) gives a much more convincing 74 unpackaged crates, but that appears to be the full dependencies-of-dependencies result.  Dropping an additional 3 where `cargo build` got errors on the debian-packaged versions but not on upstream ones did produce a `target/debug/typst`  that executes.

So there's more left than I'd hoped, but still interesting to know.
